### PR TITLE
Bypass login flow in UI tests

### PR DESCRIPTION
**Pull request by crazytonyli**
## Description

The first thing most UI tests does is sign in to a WP.com site, which is done via using XCTest to navigating the UI: type email, password, select a site, tap the Skip button on the onboarding prompt screen. This whole flow takes about 14 seconds to complete. A test case takes about 25-60 seconds to complete. That means the UI tests roughly spend 25%-50% on log in, a "set up" phase.

In this PR, I added a helper function to the app, which automatically sign in and select a test site specified by UI tests. This new feature eliminates 14 seconds off each UI test function, and in totally cut down the whole UI test suite run time from about 26 mins to 20 mins.

Since the new function checks for launch arguments that's passed from UI tests, it has no impact on app functions.

## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
None

3. What automated tests I added (or what prevented me from doing so)
None

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist: N/A

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/272">24.4</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by crazytonyli** [comment]
@wordpress-mobile/mobile-ui-testing-squad For some reason, this PR's UI Tests on iPhone consistently failed at a couple of publishing post related tests (see [the build history](https://buildkite.com/automattic/wordpress-ios/builds?branch=bypass-login-during-ui-tests)). And, I had to [change an API stub][commit] to fix the issue.

The cause of the failure was UI test can't find "View" button in "Post published" toast message which has a "View" button it. Here is my theory of the root cause.

The failed tests (`testTextPostPublish` and `testBasicPostPublishWithCategoryAndTag`) uses [the `new_post_flow` WireMock scenario][new_post_flow], whose expected state transition is "Started" -> "initial_posts_list" -> "post_published" -> "new_posts_list". The app is expected to fire a fetching post API call, which [transits the WireMock scenario state from "Started" to "initial_posts_list"][fetching_post]. And later the app is expected to fire a publish post API, whose stub [requires the scenario to be in "initial_posts_list" state][publishing_post]. However, the issue is the app for some reason didn't fire the fetching post API call, which means the scenario state is still at "Started", which causes the publishing post API stub to return a failed result.

It's possible that the auto-sign shortcut somehow makes the app skipping some code that used to be called in an normal app, but at the same time, I wonder if the "new_post_flow" scenario need to have a "initial_posts_list" state. Because [the related tests][testcase] opens an editor screen straight from the My Sites screen, which means it seems to fine to just change the expected state transition to "Started" -> "post_published" -> "new_posts_list". That's why I made [the API stub change][commit]. Please let me know whether you think that change makes sense.

[commit]: https://github.com/wordpress-mobile/WordPress-iOS/pull/22735/commits/86544f43ee975eb3f81442bddebb6bf389515850
[new_post_flow]: https://github.com/wordpress-mobile/WordPress-iOS/blob/ddd0437d751b85caf881cd42eca4231324a1cb2c/API-Mocks/WordPressMocks/src/main/assets/mocks/__files/__admin/scenarios.json#L27-L32
[fetching_post]: https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/API-Mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/posts/posts_draft%2Cpending.json#L4
[publishing_post]: https://github.com/wordpress-mobile/WordPress-iOS/blob/ddd0437d751b85caf881cd42eca4231324a1cb2c/API-Mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/posts/posts_new.json#L3
[testcase]: https://github.com/wordpress-mobile/WordPress-iOS/blob/ddd0437d751b85caf881cd42eca4231324a1cb2c/WordPress/UITests/Tests/EditorGutenbergTests.swift#L15

File: WordPress/Classes/System/WordPressAppDelegate.swift [hunk 1]
```diff
@@ -972,3 +972,29 @@ extension WordPressAppDelegate {
...
+extension WordPressAppDelegate {
+
+    func autoSignInUITestSite() {
+        let launchArguments = ProcessInfo.processInfo.arguments
+        guard let optIndex = launchArguments.firstIndex(of: "-ui-test-select-wpcom-site"),
+              optIndex + 1 < launchArguments.count else {
+            return
+        }
+
+        let wpComSiteAddress = launchArguments[optIndex + 1]
```
Comment (AliSoftware) [inline WordPress/Classes/System/WordPressAppDelegate.swift]: Tip: you can use [`UserDefaults`'s ability to auto-parse dash arguments](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/UserDefaults/AboutPreferenceDomains/AboutPreferenceDomains.html#//apple_ref/doc/uid/10000059i-CH2-96930) in [its `NSArgumentDomain`](https://developer.apple.com/documentation/foundation/nsargumentdomain) here instead, which would avoid you having to parse them yourself and check the launchArguments array boundaries and all.

```suggestion
        guard let wpComSiteAddress = UserDefaults.standard.string(forKey: "ui-test-select-wpcom-site") else {
            return
        }
```

Comment (crazytonyli) [reply WordPress/Classes/System/WordPressAppDelegate.swift]: Thanks for the tip! I've used it in https://github.com/wordpress-mobile/WordPress-iOS/pull/22735/commits/70ef2531cd9f307b666430f957d3ed0c39d580bc.

**Review by tiagomar — APPROVED** [review]
Thanks for working on this, @crazytonyli! 🙇 

The WireMock scenario change makes sense, and I see no reason not to implement it.

I've run the tests several times both individually and in bundle, in various orders and everything looks good. Since the CI is also 🟢 , it's good to go!

**Comment by crazytonyli** [comment]
@tiagomar Thanks for the review!

**Pull request merged**